django/utils/text.py

```diff
@@ -64,6 +67,76 @@ def _generator():
     return "".join(_generator())
 
 
+class DjangoHTMLParser(HTMLParser):
+    class TruncationCompleted(Exception):
+        pass
+
+    void_elements = {  # https://html.spec.whatwg.org/#void-elements
+        "area",
+        "base",
+        "br",
+        "col",
+        "embed",
+        "hr",
+        "img",
+        "input",
+        "link",
+        "meta",
+        "source",
+        "track",
+        "wbr",
+    }
+
+    def __init__(self, *, convert_charrefs=True):
+        super().__init__(convert_charrefs=convert_charrefs)
+        self.tags = deque()
+        self.output = ""
+
+    def handle_starttag(self, tag, attrs):
+        self.output += self.get_starttag_text()
+        if tag not in self.void_elements:
+            self.tags.appendleft(tag)
+
+    def handle_endtag(self, tag):
+        if tag not in self.void_elements:
+            self.output += f"</{tag}>"
+            self.tags.remove(tag)
+
+    def handle_data(self, data):
+        if self.func == "words":
+            data = re.split(r"(?<=\S)\s(?=\S)", data)
+            output = escape(" ".join(data[: self.num]))
+        else:
+            output = escape("".join(data[: self.num]))
+
+        data_len = len(data)
+        if self.num < data_len:
+            if self.func == "words" and self.truncate and self.truncate[0] == " ":
+                output = output.strip()
+            self.output += Truncator.add_truncation_text(output, self.truncate)
+            self.output += "".join([f"</{tag}>" for tag in self.tags])
+            raise self.TruncationCompleted()
+        self.num -= data_len
+        self.output += output
+
+    def _truncate_html(self, func, text, num, truncate):
+        if func not in ("words", "chars"):
+            raise AttributeError(f"func must be 'words' or 'chars', got '{func}'.")
+        if num <= 0:
+            if func == "words":
+                return ""
+            else:
+                return truncate
+        self.func = func
+        self.num = num
+        self.truncate = truncate
```

Comment:
can you move all the arguments up into `__init__` ? it's hard to read the code to see where `self.func` etc. come from at the moment, on first reading I thought they were comign from `HTMLParser`.

Reply:
See my pull request to this branch posted above. Already handles this.

> I added some suggested changes in [smithdc1#3](https://github.com/smithdc1/django/pull/3).